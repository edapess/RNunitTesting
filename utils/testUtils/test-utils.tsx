import { AppStore } from "@/app/shared/types/redux";
import { RootState, setupStore } from "@/app/store/configureStore";
import { ThemeProvider } from "@/utils/uiUtils/themeUtils";
import {
  Action,
  combineReducers,
  configureStore,
  EnhancedStore,
  Middleware,
  Reducer,
} from "@reduxjs/toolkit";
import { act, render, RenderOptions } from "@testing-library/react-native";
import React, { JSX, PropsWithChildren } from "react";
import { Provider } from "react-redux";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export async function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  function Wrapper({ children }: PropsWithChildren): JSX.Element {
    return (
      <Provider store={store}>
        <ThemeProvider>{children}</ThemeProvider>
      </Provider>
    );
  }

  const result = render(ui, { wrapper: Wrapper, ...renderOptions });

  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
  });

  return { store, ...result };
}

export function setupApiStore<
  A extends {
    reducer: Reducer<any, any>;
    reducerPath: string;
    middleware: Middleware;
    util: { resetApiState(): any };
  },
  R extends Record<string, Reducer<any, any>> = Record<never, never>,
>(api: A, extraReducers?: R): { api: any; store: EnhancedStore } {
  /*
   * modified version of RTK Query's helper function:
   * check docs for more details
   */
  const getStore = (): EnhancedStore =>
    configureStore({
      reducer: combineReducers({
        [api.reducerPath]: api.reducer,
        ...extraReducers,
      }),
      middleware: (gdm) =>
        gdm({ serializableCheck: false, immutableCheck: false }).concat(
          api.middleware,
        ),
    });

  type StoreType = EnhancedStore<
    {
      api: ReturnType<A["reducer"]>;
    } & {
      [K in keyof R]: ReturnType<R[K]>;
    },
    Action,
    ReturnType<typeof getStore> extends EnhancedStore<any, any, infer M>
      ? M
      : never
  >;

  const initialStore = getStore() as StoreType;
  const refObj = {
    api,
    store: initialStore,
  };
  const store = getStore() as StoreType;
  refObj.store = store;

  return refObj;
}
