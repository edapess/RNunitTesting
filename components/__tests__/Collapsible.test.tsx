import { renderWithProviders } from "@/utils/testUtils/test-utils";
import { fireEvent, waitFor } from "@testing-library/react-native";
import { Collapsible, collapsibleTestIDs } from "../Collapsible";

const mockedTitle = "Step 1";
const mockedContent = "This is the content of the collapsible section.";

describe("<Collapsible />", () => {
  test("Match snapshot", async () => {
    const { toJSON } = await renderWithProviders(
      <Collapsible title="mock title" />,
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
  test("Renders title correctly", async () => {
    const { getByTestId } = await renderWithProviders(
      <Collapsible title={mockedTitle} />,
    );
    await waitFor(() => {
      expect(getByTestId(collapsibleTestIDs.title.testID).props.children).toBe(
        mockedTitle,
      );
    });
  });
  test("Toggles content visibility on heading press", async () => {
    const { getByTestId, queryByText } = await renderWithProviders(
      <Collapsible title={mockedTitle}>{mockedContent}</Collapsible>,
    );
    await waitFor(() => {
      expect(queryByText(mockedContent)).toBeNull();
    });
    fireEvent.press(getByTestId(collapsibleTestIDs.heading.testID));
    await waitFor(() => {
      expect(getByTestId(collapsibleTestIDs.content.testID)).toBeTruthy();
      expect(
        getByTestId(collapsibleTestIDs.content.testID).props.children,
      ).toBe(mockedContent);
    });
  });
});
