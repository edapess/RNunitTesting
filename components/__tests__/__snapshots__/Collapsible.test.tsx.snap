// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapsible /> Match snapshot 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#ECEFF4",
      },
      undefined,
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 6,
        "opacity": 1,
      }
    }
    testID="Collapsible.heading"
  >
    <ViewManagerAdapter_SymbolModule
      animated={false}
      colors={[]}
      name="chevron.right"
      resizeMode="scaleAspectFit"
      style={
        [
          {
            "height": 24,
            "width": 24,
          },
          [
            {
              "height": 18,
              "width": 18,
            },
            {
              "transform": [
                {
                  "rotate": "0deg",
                },
              ],
            },
          ],
        ]
      }
      tint={4283192938}
      tintColor="#4C566A"
      type="monochrome"
      weight="medium"
    />
    <Text
      style={
        [
          {
            "color": "#2E3440",
          },
          undefined,
          undefined,
          {
            "fontSize": 16,
            "fontWeight": "600",
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
        ]
      }
      testID="Collapsible.title"
    >
      mock title
    </Text>
  </View>
</View>
`;
